### PR TITLE
Log PROXY protocol v2 TLVs; fix PROXY protocol parsing bugs

### PR DESCRIPTION
**Pull request by eduard-bagdasaryan**
Added %proxy_protocol::>h logformat code for logging received PROXY
protocol TLV values and passing them to adaptation services and helpers.
For simplicity, this implementation treats TLV values as text and reuses
the existing HTTP header field logging interfaces. Support for binary
TLV value logging can be added later if needed.

Also support logging of metadata extracted from the fixed portion of a
PROXY protocol header and referenced as "pseudo headers": :command,
:version, :src_addr, :dst_addr, :src_port, and :dst_port.

Also fixed several bugs in the old PROXY protocol v1/v2 parsing code:

* Buffer overrun in ConnStateData::parseProxy2p0(). The available
  local SBuf could be less than sizeof(pax), resulting in copying
  excessive bytes with SBuf::rawContent().

* Incorrect processing of malformed v1 headers lacking CRLF in
  ConnStateData::parseProxy1p0(), which waited for more data
  even if the buffer size already exceeded the maximum v1 header
  size.

* Incorrect processing of partial-buffered v1 headers, when only
  the initial (magic) part of the header has been received. The old
  code resulted with an error instead of waiting for more data in this
  case.

* Incorrect v1 header framing for proto=UNKNOWN headers.
  The code used only LF while the protocol requires CRLF for it.
    
* Do not use address information from v2 header if the header
  proto is UNSPEC.

* Incorrect v1 magic expectations (a 6-character `PROXY ` instead of the
  proper 5-character `PROXY` sequence) leading to mishandling of
  non-PROXY input. For example, receiving `PROXY\r\n` would result in
  "need more data" outcome (and long timeout) instead of an immediate
  error.

Also eliminated code duplication in HttpHeader::getByNameListMember()
and HttpHeader::getListMember(), moving the common part into a
separate getListMember() method.

Also eliminated code duplication and probably fixed a bug with applying
client_netmask parameter in ConnStateData constructor. The mask should
not be applied to localhost and IPv6 addresses but was.

Also parse PROXY protocol v2 LOCAL addresses (for logging purposes). In
compliance with the PROXY protocol specs, LOCAL addresses are still
unused for connection routing.

### CODE REVIEW

**Comment by rousskov** [comment]
I polished the PR title/description to disclose bug fixes in the title, to remove details already covered by cf.data.pre additions, to detail the client netmask bug, and to polish wording. Please review and adjust further if necessary.

File: src/proxyp/Message.h [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+class Message: public RefCountable
+{
+public:
+    typedef RefCount<Message> Pointer;
+    typedef std::vector<Two::Tlv> Tlvs;
+
+    Message(const char *ver, const uint8_t cmd = Two::cmdProxy);
+
+    /// HTTP header-like string representation of the message.
+    /// The returned string has one line per pseudo header version,
```
Comment (rousskov) [inline src/proxyp/Message.h]: I see no reason to itemize pseudo headers here, especially since some of them may not be available. Let's simplify:

```suggestion
    /// The returned string has one line per pseudo header and
    /// one line per TLV (if any).
```

... and remove the next comment line.

File: src/proxyp/Message.cc [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    command_(Two::Command(cmd)),
+    ignoreAddresses_(false)
+{}
+
+SBuf
+ProxyProtocol::Message::toMime() const
+{
+    SBufStream result;
+    for (const auto &p: PseudoHeaderFields)
+        result << p.first << ": " << getValues(p.second) << "\r\n";
```
Comment (rousskov) [inline src/proxyp/Message.cc]: I do not think we should add lines with empty pseudo header values. Please refactor to skip lines with an empty getValues() result. At that point, it may make sense to manually enumerate pseudo headers (as an optimization), so that we can skip all addresses/ports at once.

File: src/proxyp/Message.cc [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+{
+    SBufStream result;
+    char ipBuf[MAX_IPSTRLEN];
+
+    if (headerType == Two::htPseudoVersion) {
+        result << version_;
+    } else if (headerType == Two::htPseudoCommand) {
+        result << command_;
+    } else if (headerType == Two::htPseudoSrcAddr) {
+        if (!ignoreAddresses_) {
```
Comment (rousskov) [inline src/proxyp/Message.cc]: Here and below in this method, please use hasAddresses(). Unlike the low-level `ignoreAddresses_`, it relays the intent much better -- I may still want to log the addresses that others should ignore, but I clearly should not log something I do not have.
```suggestion
        if (hasAddresses()) {
```

File: src/proxyp/Message.cc [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    // cannot reuse Message::getValues(): need the original TLVs layout
+    for (const auto &tlv: tlvs)
+        result << tlv.type << ": " << tlv.value << "\r\n";
+    return result.buf();
+}
+
+SBuf
+ProxyProtocol::Message::getValues(const uint32_t headerType, const char sep) const
+{
+    SBufStream result;
```
Comment (rousskov) [inline src/proxyp/Message.cc]: Let's optimize this a little. Many of the pseudo headers do not benefit from SBufStream -- it only adds printing overheads. Declare SBufStream only in those `if` statements that need it. Each `if` statement can have its own `return`. No `else`s. In fact, I would use a `switch` (with default for TLVs) instead of `if`s.

File: src/proxyp/Message.cc [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    char ipBuf[MAX_IPSTRLEN];
```
Comment (rousskov) [inline src/proxyp/Message.cc]: Please move to the `if` statements that need it. Most do not so it makes sense to localize this ugliness.

File: src/proxyp/Message.h [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    bool ignoreAddresses_;
+};
+
+/// successful parsing result
+class Parsed
+{
+public:
+    Parsed(const Message::Pointer &parsedMessage, const size_t parsedSize):
+        message(parsedMessage),
+        size(parsedSize) { assert(bool(parsedMessage)); }
```
Comment (rousskov) [inline src/proxyp/Message.h]: Please move this implementation detail to the `.cc` file. IIRC, the callers are already there so this move is not going to make the code slower as the compiler can still inline it if needed.

File: src/proxyp/Message.h [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+    Message::Pointer message; ///< successfully parsed message; not nil
+    size_t size; ///< raw bytes parsed, including any magic/delimiters
+};
+
+typedef std::map<SBuf, Two::HeaderType> FieldMap;
+
+/// a mapping between pseudo header names and ids
+extern const FieldMap PseudoHeaderFields;
```
Comment (rousskov) [inline src/proxyp/Message.h]: I think this declaration belongs to `Elements.h`, especially since the constant is defined in `Elements.cc`. This map is not strongly related to the PROXY protocol message.

File: src/proxyp/Parser.cc [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+    Message::Pointer message = new Message("2.0", command);
+
+    if (proto == tpUnspecified || family == afUnspecified) {
+        message->ignoreAddresses();
+        // discard address block and TLVs
+    } else {
+        Parser::BinaryTokenizer tokHeader(header);
+        ParseAddresses(family, tokHeader, message);
+        // TODO: parse TLVs for local connections
```
Comment (rousskov) [inline src/proxyp/Parser.cc]:
```suggestion
        // TODO: parse TLVs for LOCAL connections
```

File: src/proxyp/Parser.cc [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+    if (proto > tpDgram)
+        throw TexcHere(ToSBuf("PROXY/2.0 error: invalid transport protocol ", proto));
+
+    const auto header = tokMessage.pstring16("header");
+
+    Message::Pointer message = new Message("2.0", command);
+
+    if (proto == tpUnspecified || family == afUnspecified) {
+        message->ignoreAddresses();
+        // discard address block and TLVs
```
Comment (rousskov) [inline src/proxyp/Parser.cc]:
```suggestion
        // discard address block and TLVs because we cannot tell
        // how to parse such addresses and where the TLVs start.
```

File: src/proxyp/Message.h [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    /// for v2 messages: the command field
+    /// for v1 messages: Two::cmdProxy
+    Two::Command command_;
+
+    /// true if the message relays no address information
+    bool ignoreAddresses_;
+};
+
+/// successful parsing result
+class Parsed
```
Comment (rousskov) [inline src/proxyp/Message.h]: Please move the Parsed class and the Parse() function declaration into proxyp/Parser.h. Some of the Message users do not need these names and there is no good reason to pollute them with these parsing details AFAICT. This move will also remove the surprise that Parse() is declared here but defined in Parser.cc.

Comment (eduard-bagdasaryan) [reply src/proxyp/Message.h]: Done at 2c067b.

**Comment by eduard-bagdasaryan** [comment]
>  I also pushed a few minor fixes and polishing commits. Please review them

Done with a minor fix at 6a9ef21.

**Comment by rousskov** [comment]
@yadij, FYI: I plan to clear this PR for merging soon.

File: src/proxyp/forward.h [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+#ifndef _SQUID_SRC_PROXYP_FORWARD_H
+#define _SQUID_SRC_PROXYP_FORWARD_H
+
+namespace ProxyProtocol
+{
+
+class Message;
+
+typedef RefCount<Message> MessagePointer;
```
Comment (yadij) [inline src/proxyp/forward.h]: Missing include of base/RefCount.h this header only passes the unit tests due to compiler optimization bugs.

File: src/proxyp/Parser.h [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_PROXYP_PARSER_H
+#define SQUID_PROXYP_PARSER_H
+
+#include "proxyp/forward.h"
+
+class SBuf;
```
Comment (yadij) [inline src/proxyp/Parser.h]: Please use sbuf/forward.h

File: src/StrList.cc [hunk 13]
```diff
@@ -131,3 +131,17 @@ strListGetItem(const String * str, char del, const char **item, int *ilen, const
     return len > 0;
 }
 
+SBuf
+getListMember(const String &list, const char *key, const char delimiter)
```
Comment (yadij) [inline src/StrList.cc]: This code was in HttpHeader because it is tied directly to HTTP parsing semantics for MiME header fields. There are parsing tolerance criteria which we may have to add to this logic to enforce certain behaviours. So please do not move it here.

The requested change to the HttpHeader:: members should remove the need for this code to be touched at all by this PR.

File: src/proxyp/Message.cc [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+#include "proxyp/Message.h"
+#include "sbuf/Stream.h"
+#include "sbuf/StringConvert.h"
+#include "SquidConfig.h"
+#include "StrList.h"
+
+ProxyProtocol::Message::Message(const SBuf &ver, const Two::Command cmd):
+    version_(ver),
+    command_(cmd),
+    ignoreAddresses_(false)
```
[the remark this answers is not in the document]
Comment (rousskov) [reply src/proxyp/Message.cc]: Please do not. Some compilers (and some humans!) dislike the mixture of inlined and in-constructor data member initializations. We should use only the former (when we can) and only use the latter (otherwise).

File: src/format/Token.h [hunk 15]
```diff
@@ -50,6 +50,9 @@ class Token
     const char *label;
     struct {
         char *string;
+        // TODO: Add ID caching for protocols other than PROXY protocol.
+        /// the cached ID of the parsed header or zero
```
[the remark this answers is not in the document]
Comment (rousskov) [reply src/format/Token.h]: FWIW, I disagree that this data member name or usage implies that the headerId stores Http::HdrType IDs. It is a header ID cache, not an HTTP header ID cache. There are many non-HTTP headers in Squid.

@yadij, if you insist, I am OK with using a more specific type for this data member (which would not be Http::HdrType) and then changing it back to uint32_t when adding caching for parsed HTTP header name IDs (per TODO). Going back and forth would be a waste of time and may increase porting headaches, but it is not a big deal. Do you insist?

File: src/client_side.cc [hunk 16]
```diff
@@ -2095,7 +1912,7 @@ ConnStateData::clientParseRequests()
 
         // try to parse the PROXY protocol header magic bytes
         if (needProxyProtocolHeader_) {
-            if (!parseProxyProtocolHeader())
+            if (!parseProxyProtocolMessage())
```
[the remark this answers is not in the document]
Comment (rousskov) [reply src/client_side.cc]: > I know a lot of Squid code uses the term "header" inconsistently and for the wrong things, but no need to follow that bad practice in the new code.

I do not think this PR uses the term "header" inconsistently and for the wrong things. You may argue otherwise, but the remainder of your change request appears to be unrelated to the above statement -- it talks about replacing "Message" with "Header" in new code, not fixing inconsistent/wrong "header" usage.


> The PROXY protocol defines a "connection header". 

The PROXY protocol specs certainly talk about a "header" at the beginning of a connection, but given the overall low quality of the PROXY protocol specs _and_ their half-baked transition from v1 (where the prefix was a lot more like a small header) to v2 (where the prefix has all the features of a full-blown message, from a "start line" to a variable-length multipart "body" with varying content types), that talk is not sufficient reason to propagate that terminology to Squid.

We decided to call "PROXY protocol header" a message because that was a better choice than the alternatives, even though the PROXY protocol specs use the term "header". It is much easier to realize (and get used to the fact) that Squid PROXY "message" is what the PROXY specs call "header", than to navigate the mess of mapping "PROXY header" to Squid APIs that use "header" for different purposes. This becomes especially clear when you think of fixed/always-there PROXY blocks such as `:command` -- they form the request line that belongs to Http::Message equivalent, not HttpHeader equivalent.

At the beginning, we were not sure whether to use "header" or "message", but I believe that subsequent development (the final result of which you see in this PR) has proven that "message" was the right choice. Its mismatch with the PROXY protocol terminology is a minor problem that is trivial to overcome. The mismatch between ProxyProtocol::Header and Http::Header (and related existing APIs) would be much uglier.

> This is very distinctly and explicitly a prefix header for the entire connection, it cannot be used like a message.

Why not?! Just because protocol X bytes are followed by protocol Y bytes, does not mean X bytes cannot be a message. For example, HTTP/1 CONNECT or Upgrade:websockets request is still a message.

> The HTTP 'header' objects which you are using internally to represent the pieces of PROXY data are formally defined by HTTP as "fields" with a field name and field value. So the types and functions for manipulating them should refer to Field, FieldName, FieldValue etc.

Where HTTP APIs are used for PROXY messages, those APIs should continue to be used with all their (poor) terminology choices. Given the volume and age of those APIs, it is much better to stay in sync with them than to fight them. I agree that HTTP APIs can be significantly improved by using field-based terminology, but if that improvement is done (and that is a big `if`!), it should be done properly across the entire code base, in a dedicated PR(s).

@yadij, I hope you do not insist on renaming ProxyProtocol::Message into ProxyProtocol::Header. Do you?

File: src/HttpHeader.h [hunk 17]
```diff
@@ -118,8 +118,8 @@ class HttpHeader
     bool hasNamed(const SBuf &s, String *value = 0) const;
     /// \deprecated use SBuf method instead.
     bool hasNamed(const char *name, unsigned int namelen, String *value = 0) const;
-    String getByNameListMember(const char *name, const char *member, const char separator) const;
-    String getListMember(Http::HdrType id, const char *member, const char separator) const;
+    SBuf getByNameListMember(const char *name, const char *member, const char separator) const;
```
[the remark this answers is not in the document]
Comment (rousskov) [reply src/HttpHeader.h]: Please do not stuff HTTP API conversion into this PR. Introducing those new APIs correctly is a large, difficult task. It is unnecessary in this PR. I do not want to end up spending days reviewing and/or rewriting those API changes as a part of this totally unrelated work. Upgrading String to SBuf alone is not a significant enough change to (legitimately) insist on the corresponding API overhaul.

File: src/format/Token.cc [hunk 18]
```diff
@@ -480,9 +487,14 @@ Format::Token::parse(const char *def, Quoting *quoting)
 
     case LFT_NOTE:
 
+    case LFT_PROXY_PROTOCOL_RECEIVED_HEADER:
+
         if (data.string) {
             char *header = data.string;
-            char *cp = strchr(header, ':');
+            // HTTP header field names cannot have ':' while
+            // PROXY protocol pseudo headers start with it.
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/format/Token.cc]: > Implementing this change request would be wrong because proxyPseudoHeader is not set to true for all pseudo headers. It is only set to true for PROXY protocol pseudo headers. The code using this variable depends on this.

Ah the lack of brackets fooled my eyes there.

> 
> It is true that the PROXY protocol specs do not use the "pseudo header" terminology (yet?), but that does not stop us from using that (general) terminology. The term matches the use case perfectly, there is no good replacement within the PROXY protocol specs, and the future HTTP/2 support would only benefit from this naming.
> 

I don't see how HTTP/2 benefits from the HTTP pseudo-headers being called PROXY-protocol pseudo-header.


> N.B. The term "pseudo header" [predates](https://en.wikipedia.org/wiki/User_Datagram_Protocol#IPv6_Pseudo_Header) HTTP/2 by many years. It is (correctly) used by HTTP/2, but it is certainly not HTTP/2-specific.

So where did Eduard get this definition of pseudo-header as a lower case string starting with a ':' if not HTTP/2 documents?

File: src/proxyp/Message.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/proxyp/Message.cc]: Lack of it makes several constructors undefined which results in move optimizations being forbidden.

File: src/ip/Address.cc [hunk 19]
```diff
@@ -101,6 +101,14 @@ Ip::Address::applyMask(Ip::Address const &mask_addr)
     return changes;
 }
 
+int
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/ip/Address.cc]: on a deeper look, it appears that this methods return result is ignored by (void) in the only caller. So removing the return result entirely is appropriate. The applyMask() return is sometimes used - as an int, so that is broken but out of scope to fix here.

File: src/proxyp/Message.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/proxyp/Message.cc]: AFAICT the modern compilers define a bitmap from the static initializers that gets blitted into the memory space on newly allocated objects, that forms the default constructor. Other constructors run that same initialization before their more-specific construct logic (if any) to fill out the memory which remains undefined or needs to be overwritten.

File: src/HttpHeader.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/HttpHeader.h]: > Upgrading String to SBuf alone is not a significant enough change to (legitimately) insist on the corresponding API overhaul.

This does not compute. This PR *does* an API overhaul, but not to the apparently desirable API. I am asking for the overhaul changes to be left out of scope and the desirable API be added for use currently only by this new caller.

File: src/ip/Address.cc (hunk 19 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/ip/Address.cc]: I think we have a miscommunication here.

What I mean is that this new applyClientMask() method does not need to produce a return value, so do not produce one. Just call the relevant (void)applyMask(mask); and have no return values from applyClientMask().

File: src/client_side.cc (hunk 16 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/client_side.cc]: Amos:
>  I know a lot of Squid code uses the term "header" inconsistently and for the wrong things, but no need to follow that bad practice in the new code.

Alex:
> I do not think this PR uses the term "header" inconsistently and for the wrong things. You may argue otherwise, but the remainder of your change request appears to be unrelated to the above statement -- it talks about replacing "Message" with "Header" in new code, not fixing inconsistent/wrong "header" usage.

Other code uses "header" to mean both the mime area of messages and an individual field in that. This PR currently re-names what the protocol specification defines as a "header" to be "message" and field-names to "headers". Even HTTP does not use that combination of definitions - a "message" there is a request-line or status-line and 0-N segments/frames of other stuff.
How can you say that this PR is not perpetuating the bad / vague naming practice?


Amos:
> This is very distinctly and explicitly a prefix header for the entire connection, it cannot be used like a message.

Alex:
> Why not?! Just because protocol X bytes are followed by protocol Y bytes, does not mean X bytes cannot be a message. For example, HTTP/1 CONNECT or Upgrade:websockets request is still a message.

This is exactly the type of confused conclusion that results from thinking of the PROXY header as a message. The bytes forming a PROXY protocol header cannot and must not be delivered after any previous bytes in that same connection. It is defining the properties of the native protocol(s) within which it is wrapped.

Eduard:
> a developer, unfamiliar with PROXY protocol, could mistakenly expect that sometimes there may be PROXY bytes beyond this 'header'. Naming all PROXY-protocol-related bytes as 'message' resolves such confusion.

Firstly this object is a MiME representation of the PROXY protocol. Mime headers do not have anything "following". They are by definition a list of things: the end of the list is the end of the list, with no such concept as "the list entry *after* the end of the list".

Secondly, a sane developer would look for the specifications of these bjects. Especially since we should be referencing the specification the documents in relevant object documentation. If one searches for the terms used in this PR in either the PROXY or HTTP specifications one finds text describing "message" as a thing - that text clearly indicates messages as things with multiple and repeated uses - with conflicts in a critical way with the actual PROXY definition for use of the "header" bytes here being called a "Message".

This thing which is only defined here and being called "Message" is representing the PROXY octets in the form of a "MiME header". Which is formally defined in the MiME specification (re-used by HTTP, etc) as a "set of header fields" (note that header here is part of a compound noun, much like "transparent" the second word actually has more specific meaning for concept clarity).

In short - I do not see any confusion if we refer to the field-name:field-value pairs as fields, field-name, and field-value. Those terms are consistent with the HTTP, MiME and PROXY specifications terminology for and thus should be relatively familiar to anyone. Likewise "Header" for a set of such fields is also consistent for all those specifications (if a bit loose).

File: src/format/Token.cc (hunk 18 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/format/Token.cc]: So essentially you took inspiration and based a pseudo-header definition on the HTTP/2 design. As I said.

The problem I have with this code is that a) the "pseudo headers" is your own invention, but the documentation makes it sound like they are a thing defined by PROXY protocol. And b) that the variable name reinforces that false implication.

We do not actually gain anything by mentioning PROXY protocol here. We *loose* the ability to log pseudo-header fields by preventing them generically. While also forcing extra work on dev trying to add other protocols which share this ":name" pseudo-header design (aka, HTTP/2 and HTTP/3). What one would call a net-negative.

File: src/proxyp/Message.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/proxyp/Message.cc]: > Lack of it makes several constructors undefined which results in move optimizations being forbidden.

I am not sure why you mention this -- if doing X leads to compiler warnings, then any positive side effects of doing X are irrelevant in Squid context.

As for other constructors being left "undefined", it is good that they are automatically deleted/forbidden for this class because the class objects should not be copied (at least not without extra work that is better spent elsewhere).

> I have not found examples restricting their usage.

I suspect the compiler warnings in question are due to a compiler flaw/deficiency (i.e., it does not support something that it should support) and/or due to compiler authors belief that mixing initialization styles is either dangerous or disables something useful (i.e., it is just a compiler-specific warning, not a language-required prohibition). Unfortunately, I do not have a warning example to use for looking the warning explanation up in the compiler documentation.

It is also possible that my recollection about these warnings/restrictions is faulty. I believe Amos was the first who discovered this issue (in Squid context), but I do not have a ready-to-use reference to that email thread. I do not remember whether I have seen this warning myself, but it is likely that I have not.

@yadij, if this change request is no longer relevant, please mark this thread accordingly by clicking the "Resolve conversation" button. Otherwise, please post a conversation comment to clarify why the discussion so far has not addressed the change request.

File: src/HttpHeader.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/HttpHeader.h]: > This PR _does_ an API overhaul

I assuming we are still talking about the HttpHeader API. No, this PR does not overhaul that API. It changes the return type of a few methods from one string type to another string type. That is not an API overhaul. That is a small/minor optimization. This string type replacement is not a sufficient ground to insist on adding some new API methods.

File: src/format/Token.cc (hunk 18 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/format/Token.cc]: @eduard-bagdasaryan, I suggest applying the changes @yadij insists on in this change request, simply to avoid wasting more time on this minor issue:

* The source comment content is just not that important. Amos sees a minor implication that you think is not there. That argument is nearly impossible to resolve using logic, even if all sides are perfectly reasonable (perfectly reasonable folks would not have such arguments to start with, but nobody is perfect).
* As for the "restriction" that the proxyPseudoHeader variable is only true for PROXY protocol, I suggest removing the `LFT_PROXY_PROTOCOL_RECEIVED_HEADER` condition from the variable initialization code and renaming the variable to pseudoHeader. To avoid creating a bug where something like `:Accept-Encoding` is misinterpreted as an HTTP `Accept-Encoding` header, add another `if` statement to throw an error if a pseudo header is used for the `type` value other than `LFT_PROXY_PROTOCOL_RECEIVED_HEADER`. These modifications would make the new pseudo header configuration support a bit safer AFAICT.

Hopefully, the above adjustments will address this change request.

File: src/format/Token.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/format/Token.h]: @eduard-bagdasaryan, I recommend doing what @yadij insists on in this change request. Using specific ID type for a generic ID cache (that is only used for one ID type right now) is not wrong enough to waste more time on. Whoever adds support for HTTP header type caching here would be able to easily fix this.

File: src/format/Token.cc (hunk 18 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/format/Token.cc]: This change broke parsing of integer-based TLV value types. I have a fix at https://github.com/measurement-factory/squid/commit/39b68165ad34dd95647e0caaa990aa23254cd43b

While fixing that bug, I also notice that our detection and diagnostics for certain likely configuration mistakes is sloppy and added https://github.com/measurement-factory/squid/commit/98adf59e597f95969fbac1c36c9a68fd46cf653f

Please review/retest/adjust and integrate into this PR.

**Comment by yadij** [comment]
> * {023} still gets a successful value of 23


Number strings that start with '0' are Octal numbers, base-7. So "023" should produce a value of 17.

**Comment by eduard-bagdasaryan** [comment]
> this PR should instead prohibit multi-digit numbers that start with 0

Done at c11949.

File: src/proxyp/Elements.cc [hunk 20]
```diff
@@ -50,7 +50,8 @@ ProxyProtocol::IntegerToHeaderType(const SBuf &rawInteger)
     int64_t tlvType = 0;
 
     Parser::Tokenizer ptok(rawInteger);
-    if (!ptok.int64(tlvType, 10, false) || !ptok.atEnd())
+    const auto parsed = ptok.skip('0') || ptok.int64(tlvType, 10, false);
+    if (!parsed || !ptok.atEnd())
```
Comment (rousskov) [inline src/proxyp/Elements.cc]: This new code works (AFAICT), but it is very difficult to comprehend, and, from a leading zeros _writer_ pointer of view, it produces a rather strange exception -- such a writer is nearly guaranteed to think that `010` _is_ a decimal integer. Please refactor to better target leading zeros. Here is a sketch:

```
if (ptok.skip(0)) {
    if (!ptok.atEnd())
        throw TexcHere(ToSBuf("Invalid PROXY protocol TLV type value. ",
                              "Expected a decimal integer without leading zeros but got '",
                              rawInteger, "'")); // e.g., 077, 0xFF, or 0b101
    // tlvType stays zero
} else {
    Must(ptok.int64(tlvType, 10, false)); // the first character is a DIGIT
    if (!ptok.atEnd())
        throw TexcHere(ToSBuf("Invalid PROXY protocol TLV type value. ",
                              "Trailing garbage after ", tlvType, " in '",
                              rawInteger, "'")); // e.g., 1.0 or 5e0
}
```

Comment (eduard-bagdasaryan) [reply src/proxyp/Elements.cc]: > Please refactor to better target leading zeros.

Done at d9f9458.

File: src/format/Token.cc (hunk 18 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/format/Token.cc]: For the record: Eduard has added both fixes as of branch commit 878714dce37300225b833b719e5f072143183e15. Determining whether this particular change request has been addressed remains on @yadij side.

File: src/proxyp/Message.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/proxyp/Message.cc]: FWIW, looking at it in far more detail I found another members' sub-member type was not providing C++11 semantics yet either so this change alone would not be sufficient to enable the constructor optimizations anyway.

I still think we could do better by always using C++11 initialization - even when there are some members needing custom construct. But is not worth blocking progress at this point.

File: src/HttpHeader.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/HttpHeader.h]: @Alex, the internals of these getByNameListMember and getListMember functions has been completely re-written using new data types and new logic sequences. So there is no less review needed for the current approach. In fact the opposite..

What I am disliking is that by re-using the old member names this PR adds a review of all the old callers to the workload. If this PR were only shuffling the code around, or using the old member signatures - no problem. But a full re-write of the member both internally and symbol types is significant change. Either we agree to keep the "Foo get*(...)" style of members, or we push the switch to your "bool has*()" design whenever anyone touches the class methods (as done here).

File: src/client_side.cc (hunk 16 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/client_side.cc]: I do. On grounds that the PROXY spec document uses the term "header" as its definition for the protocol object being represented. If you want to discuss altering that naming with Willy T. go ahead, the document is technically still in draft and he is very open to improvements if we have issues.

I just want the terminology we use in the code symbol names to align with the specification texts that code derives from so we don't have to re-educate people on why things are "misnamed" or impose a particular view of the world as necessary to even read the code. We have enough problems in that regards with the English language dialects barrier between ourselves.

File: src/StrList.cc (hunk 13 above)
Comment (yadij) [reply src/StrList.cc]: I am asking because this header is on the RoadMap for removal and all the code in it is going to be the focus for removal, some in ongoing branches, some later. If you can avoid adding things here it will help reduce that workload. This is not a blocker though, just a request.

File: src/client_side.cc (hunk 16 above)
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/client_side.cc]: > I do. On grounds that the PROXY spec document uses the term "header"

Done at b966eb and a25f68.

> Do you want s/HeaderType/FieldType/i in ProxyProtocol code? I am OK with that change.

Done at 35898f.

File: src/HttpHeader.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/HttpHeader.h]: > getListMember functions has been completely re-written using new data types and new logic sequences.

It may seem so but is not the reality. I see one 'logic' change: the new getListMember() got additional safety array-boundary check `static_cast<size_t>(ilen) > keyLen`. Other logic is the same. 

> review of all the old callers

If you'd examined these callers, you probably noticed that:
* some of the old callers naturally required SBuf though (format/Format.cc), using the clumsy StringToSBuf conversion. The new code there (which should be in line with the similar calls) would require these conversions either, which is not perfect.
* some of the old callers only cared about the emptiness of the obtained string. Is it really important what method we call there, String::size() or SBuf::isEmpty()?

[the remark this answers is not in the document]
Comment (rousskov) [reply src/HttpHeader.h]: > So there is no less review needed for the current approach. In fact the opposite..

That is just your opinion about the time _I_ will have to spend. In reality, the effort I usually spend on reviewing or writing APIs is often far greater than the effort spent on reviewing method implementations, especially basic ones. And I already reviewed those implementations, so we are comparing a significant API writing effort with zero.

> Either we agree to keep the "Foo get*(...)" style of members, or we push the switch to your "bool has*()" design whenever anyone touches the class methods (as done here).

You have no right to impose such rigid restrictions and false dichotomies.

If you insist on rewriting these APIs in this PR, then I suggest reverting the corresponding upgrade to SBuf (since you are abusing that upgrade to demand out-of-scope API changes). I hope you do not insist, especially after this PR fixed so many bugs, but I was also hoping you would not even _start_ this conversation, so I am prepared to be proven wrong.

@yadij, do you insist on rewriting these APIs in this PR?

**Comment by rousskov** [comment]
It looks like my last change request has been addressed, thank you. The ball is on @yadij side AFAICT.

File: src/StrList.cc (hunk 13 above)
Comment (yadij) [reply src/StrList.cc]: resolving this issue since it is wishlist.

File: src/HttpHeader.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/HttpHeader.h]: I am sorry that you went and put a lot of work into a review that IMO should have only gone to the overview stage before opting to do the smaller code changes. Resolving this issue now rather than getting into yet another idealogical discussion.

Having the API use SBuf is a good long=term result anyway, it just means that this approach of changing so many callers implicit behaviour makes it to wide-spread to risk backporting for v4 stable series. Which is unfortunate given the PROXYv2 issues this fixes.

[the remark this answers is not in the document]
Comment (rousskov) [reply src/HttpHeader.h]: > Having the API use SBuf [...] means that this approach of changing so many callers implicit behaviour makes it to wide-spread to risk backporting for v4 stable series.

If you are worried about any callers that did not get explicitly changed, then it is easy to expose all of them during a v4 port -- a temporary rename of the affected functions will automatically expose all callers in a full build. I do that sort of checks a lot, and they are easy to apply and revert. If you want Factory to perform this check for your v4 port, just ask. We can also prepare a v4-specific PR.
